### PR TITLE
[Flink-32356][doc] Add doc for procedure

### PR DESCRIPTION
**Pull request by luoyuxia**
## What is the purpose of the change
To add the doc for call procedures

## Brief change log
- Add the doc for how to develop a procedure
- Add the doc for how to call the procedure

## Verifying this change
- N/A

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: docs/content/docs/dev/table/procedures.md [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+The class must be declared `public`, not `abstract`, and should be globally accessible. Thus, non-static inner or anonymous classes are not allowed.
+
+### Call Methods
+
+The interface doesn't provide any method，you have to define a method named `call` in which you can implement the logic of the procedure.
+The methods must be declared `public` and take a well-defined set of arguments. 
+
+Please note:
+
+* The first parameter of the method `call` should always be `ProcedureContext` which provides the method `getExecutionEnvironment` to get a `StreamExecutionEnvironment` to enable to run a Flink Job
```
Comment (hackergin) [inline docs/content/docs/dev/table/procedures.md:50]:
```suggestion
* The first parameter of the method `call` should always be `ProcedureContext` which provides the method `getExecutionEnvironment` to get a `StreamExecutionEnvironment` for running a Flink Job
```

File: docs/content.zh/docs/dev/table/procedures.md [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+)
+@ProcedureHint(
+  input = Array(),
+  output = new DataTypeHint("BOOLEAN")
+)
+class OverloadedProcedure extends Procedure {
+
+  // 一个 call 方法的实现，确保 call 方法存在于存储过程类中，可以被 JVM 调用
+  @varargs
+  def eval(context: ProcedureContext, o: AnyRef*): Array[AnyRef]= {
```
Comment (hackergin) [inline docs/content.zh/docs/dev/table/procedures.md:329]: call ?

File: docs/content/docs/dev/table/sql/call.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+specific language governing permissions and limitations
+under the License.
+-->
+
+# Call Statements
+
+`Call` statements are used to call a stored procedure which is usually provided to perform data manipulation or administrative tasks.
+
+<span class="label label-danger">Attention</span> Currently, `Call` statements require the procedure called to exist in the corresponding catalog. So, please make sure the procedure exists in the catalog. 
+If it doesn't exist, it'll throw an exception. You may need to refer to the doc of the catalog to see the available procedures. To implement an procedure, please refer to [Procedure]({{< ref "docs/dev/table/procedures" >}}).
```
Comment (hackergin) [inline docs/content/docs/dev/table/sql/call.md:32]:
```suggestion
If the procedure does not exist, an exception will be thrown. You may need to refer to the doc of the catalog to see the available procedures. To implement a procedure, please refer to [Procedure]({{< ref "docs/dev/table/procedures" >}}).
```

File: docs/content/docs/dev/table/sql/call.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+## Run a CALL statement
+
+{{< tabs "call statement" >}}
+
+{{< tab "Java" >}}
+
+CALL statements can be executed with the `executeSql()` method of the `TableEnvironment`. The `executeSql()` will call the procedure immediately, and return a `TableResult` instance which associates the procedure.
```
Comment (hackergin) [inline docs/content/docs/dev/table/sql/call.md:40]:
```suggestion
CALL statements can be executed with the `executeSql()` method of the `TableEnvironment`. The `executeSql()` will immediately call the procedure and return a `TableResult` instance which associates the procedure.
```

File: docs/content/docs/dev/table/sql/call.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+The following examples show how to run a CALL statement in `TableEnvironment`.
```
Comment (hackergin) [inline docs/content/docs/dev/table/sql/call.md:42]:
```suggestion
The following examples show how to execute a CALL statement in `TableEnvironment`.
```

File: docs/content/docs/dev/table/sql/call.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+{{< /tab >}}
+{{< tab "Scala" >}}
+
+CALL statements can be executed with the `executeSql()` method of the `TableEnvironment`. The `executeSql()` will call the procedure immediately, and return a `TableResult` instance which associates the procedure.
```
Comment (hackergin) [inline docs/content/docs/dev/table/sql/call.md:47]:
```suggestion
CALL statements can be executed with the `executeSql()` method of the `TableEnvironment`. The `executeSql()` will immediately call the procedure and return a `TableResult` instance which associates the procedure.
```

File: docs/content/docs/dev/table/sql/call.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+The following examples show how to run a single CALL statement in `TableEnvironment`.
+{{< /tab >}}
+{{< tab "Python" >}}
+
+CALL statements can be executed with the `execute_sql()` method of the `TableEnvironment`. The `executeSql()` will call the procedure immediately, and return a `TableResult` instance which associates the procedure.
```
Comment (hackergin) [inline docs/content/docs/dev/table/sql/call.md:53]:
```suggestion
CALL statements can be executed with the `execute_sql()` method of the `TableEnvironment`. The `executeSql()` will immediately call the procedure and return a `TableResult` instance which associates the procedure.
```

File: docs/content/docs/dev/table/sql/call.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+The following examples show how to run a single CALL statement in `TableEnvironment`.
```
Comment (hackergin) [inline docs/content/docs/dev/table/sql/call.md:55]:
```suggestion
The following examples show how to execute a single CALL statement in `TableEnvironment`.
```

File: docs/content/docs/dev/table/sql/call.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+{{< /tab >}}
+{{< tab "SQL CLI" >}}
+
+CALL statements can be executed in [SQL CLI]({{< ref "docs/dev/table/sqlClient" >}}).
+
+The following examples show how to run a CALL statement in SQL CLI.
```
Comment (hackergin) [inline docs/content/docs/dev/table/sql/call.md:62]:
```suggestion
The following examples show how to execute a CALL statement in SQL CLI.
```

File: docs/content/docs/dev/table/procedures.md [hunk 10]
```diff
@@ -0,0 +1,544 @@
...
+
+Flink's procedures implement an automatic type inference extraction that derives data types from the procedure's class and its `call` methods via reflection. If this implicit reflective extraction approach is not successful, the extraction process can be supported by annotating affected parameters, classes, or methods with `@DataTypeHint` and `@ProcedureHint`. More examples on how to annotate procedures are shown below.
+
+Note: although the return type in `call` method must be array type `T[]`, if use `@DataTypeHint` to annotate the return type , it's actually expected to annotate the component type of the array type, which is actually `T`.
+
+#### Automatic Type Inference
+
+The automatic type inference inspects the procedure's class and `call` methods to derive data types for the arguments and result of a procedure. `@DataTypeHint` and `@ProcedureHint` annotations support the automatic extraction.
+
+For a full list of classes that can be implicitly mapped to a data type, see the [data type extraction section]({{< ref "docs/dev/table/types" >}}#data-type-extraction).
```
Comment (hackergin) [inline docs/content/docs/dev/table/procedures.md:142]:
```suggestion
For a full list of classes that can be implicitly mapped to a data type, please refer to the [data type extraction section]({{< ref "docs/dev/table/types" >}}#data-type-extraction).
```

File: docs/content/docs/dev/table/procedures.md [hunk 11]
```diff
@@ -0,0 +1,544 @@
...
+  }
+}
+
+```
+{{< /tab >}}
+{{< /tabs >}}
+
+
+### Return Procedure in Catalog
+After implement a procedure, the catalog can then return the procedure in method `Catalog.getProcedure(ObjectPath procedurePath)`. The following example shows how to return it in a catalog.
```
Comment (hackergin) [inline docs/content/docs/dev/table/procedures.md:344]:
```suggestion
After implementing a procedure, the catalog can then return the procedure in method `Catalog.getProcedure(ObjectPath procedurePath)`. The following example shows how to return it in a catalog.
```

**Review by hackergin — COMMENTED** [review]
@luoyuxia Thanks for your contribution, I left some comments, PTAL.

File: docs/content.zh/docs/dev/table/sql/call.md [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+{{< /tabs >}}
+
+{{< top >}}
+
+## Syntax
+
+```sql
+CALL [catalog_name.][database_name.]procedure_name ([ expression [, expression]* ] )
+```
+
```
Comment (hackergin) [inline docs/content.zh/docs/dev/table/sql/call.md:118]: We can add a part about  list all procedure here.

**Comment by hackergin** [comment]
The title of the PR should be changed to  [Flink-32356][doc] Add doc for procedure

File: docs/content.zh/docs/dev/table/procedures.md [hunk 13]
```diff
@@ -0,0 +1,541 @@
...
+```
+{{< /tab >}}
+{{< tab "Scala" >}}
+```scala
+import org.apache.flink.table.procedure.ProcedureContext
+import org.apache.flink.table.procedures.Procedure
+import scala.annotation.varargs
+
+// 有多个重载 call 方法的存储过程
+class GenerateSequenceProcedure extends Procedure {
```
Comment (hackergin) [inline docs/content.zh/docs/dev/table/procedures.md:106]:
```suggestion
class GenerateSequenceProcedure implements Procedure {
```

Comment (luoyuxia) [reply docs/content.zh/docs/dev/table/procedures.md:106]: In scala, it should be `extends`.

File: docs/content.zh/docs/dev/table/sql/call.md (hunk 12 above)
Comment (luoyuxia) [reply docs/content.zh/docs/dev/table/sql/call.md:118]: It has been added before. See https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/show/#show-procedures

**Comment by luoyuxia** [comment]
@hackergin Thanks for review. Merging....

**Pull request closed without merging**